DX-902: Add a non-ASCII guard with documentation and tests

azeinela opened the pull request:
## Summary
- add a security-focused `detect-non-ascii-characters` hook to prevent prompt injection attacks via malicious Unicode content
- implement three configurable modes: `balanced` (default, allows ASCII + Latin-1), `visible-plus` (allows ASCII + emoji), and `ascii-only` (strict)
- add grapheme-aware scanning using `grapheme>=0.6.0` to properly handle emoji clusters and ZWJ sequences
- support flexible filtering via `--include-range`, `--allow-chars`, `--files-glob`, `--files-include`, `--files-exclude`, and `--check-only` options
- document the new hook with usage examples and register its console script entry

Commit:
Add a non-ASCII guard with documentation and tests

Commit:
Fix newline at end of files

Commit:
use Windows-safe byte literals in non-ASCII guard test

Commit:
Refactor main function to always write new data and increase coverage to 100%

rzuckerm reviewed (changes requested):
We need to discuss this when Trevor gets back

Commit:
Enhance non-ASCII guard with new modes and file filtering options; update tests accordingly

Commit:
set encoding for tests

Commit:
increase coverage

Commit:
add more tests for coverage

rzuckerm commented:
This PR needs to be made against the `adtran` branch, not `main`.

trevorbowen reviewed (approved):
lgtm, although I'd like to see some comments in the code that detail in plain english what each of the ASCII values are, like:

```
# ASCII values correspond to:
#   0x09 - horizontal tab, \t
#   0x0A - carriage return, \r
#   0x0D - newline, \n
#   0x20-0x7E - printable characters: letters (upper, lower), digits (0-9), punctuation (!, @, #, $. %, ^, &), and space
```

You could break it down more, if you like, but my main concern was that I knew some special values are control characters, which should not be allowed, but I could not remember what they were without looking them up.

---

A demo would be nice too.  :smile:

rzuckerm commented on README.md:
```diff
@@ -111,6 +111,23 @@ The following arguments are available:
+    - `ascii-only`: allow only tab/lf/cr and `0x20-0x7E`; block everything else.
+    - Examples: `--mode balanced` (default); `--mode visible-plus` (allow 😀, 🚀, etc. but still block zero-width joiners); `--mode ascii-only` (paranoid mode, blocks all non-ASCII).
+  - `--include-range RANGE` - override allowed byte ranges (comma-separated, decimal or hex, supports `START-END`). Can be repeated.
+    - Examples: `--include-range 0x09,0x0A,0x0D,0x20-0x7E` (default printable ASCII); `--include-range 0-255` (allow all bytes); `--include-range 0x20-0x7E,0xA0` (allow NBSP too).
```
I'm assuming that `0-255` means Unicode 0 to 255. Otherwise, this would basically disable this checker.

rzuckerm commented on pre_commit_hooks/non_ascii_guard.py:
```diff
@@ -0,0 +1,267 @@
+        print(
+            f'Fixing {filename}: ' f'disallowed bytes {_format_offenders(offenders)}',
+        )
+        retv = 1
```
Should this be removed since this is not the `check_only` case?

rzuckerm commented on setup.cfg:
```diff
@@ -18,7 +18,11 @@ classifiers =
+install_requires =
+    ruamel.yaml>=0.15
+    grapheme>=0.6.0
+    tomli>=1.1.0;python_version<"3.11"
```
This needs to be removed since that is handled in the `requirements-dev.txt`.

rzuckerm commented on pre_commit_hooks/non_ascii_guard.py:
```diff
@@ -0,0 +1,267 @@
+            ch = chr(b)
+            # use repr to surface escapes for backslash/quote while staying ASCII
+            ch_repr = repr(ch)[1:-1]
+            return f"0x{b:02x}('{ch_repr}')@{pos}"
```
Showing the byte number does not seem like the best way to handle this. It seems like there should be a line number and character number instead.

rzuckerm commented on pre_commit_hooks/non_ascii_guard.py:
```diff
@@ -0,0 +1,267 @@
+            continue
+
+        if args.check_only:
+            print(f'{filename}: disallowed bytes {_format_offenders(offenders)}')
```
Should this be shown regardless of `check_only`?

rzuckerm commented on tests/non_ascii_guard_test.py:
```diff
@@ -0,0 +1,326 @@
+
+
+def test_visible_plus_blocks_bidi_in_cluster(tmp_path, capsys):
+    """Test line 107: bidi override check in _cluster_allowed_visible_plus"""
```
Please either remove this docstring. It is unnecessary, and the line number will change when the code changes. The same goes for all other tests with docstrings.

Commit:
Refactor non-ASCII character detection and update tests

- Removed the non_ascii_guard.py script and replaced it with detect_non_ascii_characters.py for improved clarity and functionality.
- Updated setup.cfg to reflect the new script name and versioning.
- Added new tests for the non-ASCII character detection functionality, ensuring comprehensive coverage of various scenarios including mode checks, file filtering, and character allowances.
- Removed outdated tests related to the old non_ascii_guard implementation.
- Updated requirements-dev.txt to include the new grapheme-cluster-break dependency.

Commit:
fix code coverage

Commit:
remove bad tests

Commit:
Refactor error handling in tests and update file writing methods for consistency

rzuckerm commented on pre_commit_hooks/detect_non_ascii_characters.py:
```diff
@@ -0,0 +1,546 @@
+ASCII_BASE = {0x09, 0x0A, 0x0D} | set(range(0x20, 0x7F))  # tab/LF/CR + space-tilde (printable ASCII)
+LATIN1_VISIBLE = set(range(0xA0, 0x100))  # Latin-1 Supplement: é, ñ, ç, etc.
+LATIN_EXT_A = set(range(0x0100, 0x0180))  # Latin Extended-A: Polish, Czech, Hungarian, etc.
+CONTROL_C0 = set(range(0x00, 0x20))  # C0 controls (except tab, LF, CR which are in ASCII_BASE)
```
Should this remove tab, LF, and CR? The comment implies this

azeinela replied:
These are permitted invisible characters alongside the ascii base.

rzuckerm replied:
Seems like this should be this, or the comment is wrong:

```suggestion
CONTROL_C0 = set(range(0x00, 0x20)) - ALLOWED_WHITESPACE  # C0 controls (except tab, LF, CR which are in ASCII_BASE)
```

rzuckerm commented on pre_commit_hooks/detect_non_ascii_characters.py:
```diff
@@ -0,0 +1,546 @@
+        _cluster_allowed_balanced(cps) if mode == MODE_BALANCED else
+        _cluster_allowed_visible_plus(cps) if mode == MODE_VISIBLE_PLUS else
+        _cluster_allowed_ascii_only(cps)
+    )
```
This would be clearer like this:

```suggestion
    if mode == MODE_BALANCED:
        _cluster_allowed_balanced(cps)
    elif mode == MODE_VISIBLE_PLUS:
        _cluster_allowed_visible_plus(cps)
    else:
        _cluster_allowed_ascii_only(cps)
```

rzuckerm reviewed (changes requested):
Overall this looks good. However, there are some coding practice that need to be changed:
- Avoid doing this:
  ```python
  if condition: statement
  ```
  do this instead:
  ```python
  if condition:
      statement
  ```
  That way it is easier to add more statements if needed in the future
- Similarly avoid doing this:
  ```python
   except <exception>:  statement
  ```
  do this instead:
  ```python
  except <expection>:
      statement
  ```
  Same reason as above
- There seem to be several `# pragma: no cover` comments. It seems like these paths could be covered if you added the necessary patches in your tests:
  ```python
  from unittest.mock import patch

  @patch("pre_commit_hooks.detect_non_ascii_character.<path-to-function-being-patch>")
  def test_something(mock_whatever):
       mock_whatever.return_value = <some-value>
       # or
       mock_whatever.side_effect = <some-exception>("some message")
  ```
  See [this](https://docs.python.org/3.9/library/unittest.mock.html) for details.

rzuckerm commented:
Other comments:
- Avoid doing this:
  ```python
  a, b, c = value1, value2, value
  ```
  Do this instead:
  ```
  a = value1
  b = value2
  c = value3
  ```
- There are a lot of magic number in the code. Please add named constants for these.

Commit:
improved readability, coverage, progamming best practices

azeinela commented:
### Summary
- Security-focused hook to prevent prompt injection attacks via malicious Unicode (invisible chars, bidi overrides, control sequences)
- Three flexible modes: balanced (ASCII + Latin-1), visible-plus (ASCII + emoji), ascii-only (strict)
- Grapheme-aware scanning with proper emoji/ZWJ sequence handling and intelligent binary file detection
- Auto-fix capability with NFKD (Normalization Form KD) normalization to decompose accented characters in ascii-only mode
-  Flexible filtering via --include-range, --allow-chars, --file-include, --file-exclude options with conflict detection

Commit:
fix windows os test coverage

rzuckerm commented on pre_commit_hooks/detect_non_ascii_characters.py:
```diff
@@ -137,15 +163,19 @@ def parse_gitattributes(path: str) -> dict:
+                    if attr.startswith('filter=lfs'):
+                        attrmap['filter'] = 'lfs'
+                    elif attr in ('binary', '-text'):
+                        attrmap['binary'] = True
```
These are not mutually exclusive. For example:
```
*.tgz filter=lfs diff=lfs merge=lfs -text
```

```suggestion
                    if attr.startswith('filter=lfs'):
                        attrmap['filter'] = 'lfs'
                    if attr in ('binary', '-text'):
                        attrmap['binary'] = True
```

rzuckerm commented on pre_commit_hooks/detect_non_ascii_characters.py:
```diff
@@ -0,0 +1,586 @@
+#   0x61-0x7A - lowercase: a-z
+#   0x7B-0x7E - punctuation: { | } ~
+
+ASCII_BASE = {0x09, 0x0A, 0x0D} | set(range(0x20, 0x7F))  # tab,LF,CR,space-tilde + printable ASCII
```
This first set is already assigned to a variable.
```suggestion
ASCII_BASE = ALLOWED_WHITESPACE | set(range(0x20, 0x7F))  # tab,LF,CR,space-tilde + printable ASCII
```

rzuckerm commented on pre_commit_hooks/detect_non_ascii_characters.py:
```diff
@@ -0,0 +1,586 @@
+    """Check if file is marked as binary in .gitattributes."""
+    for pattern, attrs in gitattributes.items():
+        if fnmatch.fnmatch(filename, pattern):
+            if attrs.get('filter') == 'lfs' or attrs.get('binary') or attrs.get('text') == False:
```
The `text` key in `attrs` is never set:

```suggestion
            if attrs.get('filter') == 'lfs' or attrs.get('binary'):
```

Commit:
Refactor code structure for improved readability and maintainability

rzuckerm commented on requirements-dev.txt:
```diff
@@ -5,3 +5,4 @@ pytest
 # Removed from setup.cfg
 ruamel.yaml>=0.15
 tomli>=1.1.0;python_version<"3.11"
+grapheme-cluster-break>=1.1.1
```
Missing dependency:

```suggestion
grapheme-cluster-break>=1.1.1
binaryornot>=0.4.4
```

rzuckerm commented on pre_commit_hooks/detect_non_ascii_characters.py:
```diff
@@ -0,0 +1,760 @@
+    if glob_conflicts:
+        return (
+            f"Conflicting file filters: --file-include and --file-exclude both use "
+            f'"{sorted(glob_conflicts)[0]}". This would exclude all files.'
```
`sorted(x)[0]` is equivalent to `min(x)`:

```suggestion
            f'"{min(glob_conflicts)}". This would exclude all files.'
```

rzuckerm commented on tests/detect_non_ascii_characters_test.py:
```diff
@@ -0,0 +1,686 @@
+import pytest
+import argparse
+from pathlib import Path
+from pre_commit_hooks import detect_non_ascii_characters as dna_hook
```
🧬 🤣

rzuckerm commented on tests/detect_non_ascii_characters_test.py:
```diff
@@ -0,0 +1,686 @@
+            assert k in attrs[key]
+            assert attrs[key][k] == v
+    for key in attrs:
+        assert key in expected_attrs
```
This seems like an overly complicated way to compare two `dict`s. Can't you just do this?

```suggestion
    assert attrs == expected_attrs
```

rzuckerm reviewed (changes requested):
A lot of the tests have a large number of parameters. Consider use a single parameter that is either a `dict` or a dataclass. Also, please use `pytest.param` with an `id`. This will make the tests easier to troubleshoot if tests fail.

Commit:
Refactor code structure for improved readability and maintainability

rzuckerm reviewed (approved):
LGTM, thanks for your diligence on this @azeinela !

Commit:
update tests to be compatible with Windows

Commit:
Fix re.sub escape issue for Windows paths in error truncation

Commit:
Remove unused testing resource

Commit:
Refactor binary file detection to utilize git attributes and improve test coverage

rzuckerm commented on pre_commit_hooks/detect_non_ascii_characters.py:
```diff
@@ -0,0 +1,747 @@
+            return result
+
+        check_attr = subprocess.run(  # Check attributes for all tracked files to find those with filter=lfs or binary
+            ("git", "check-attr", "filter", "binary", "-z", "--stdin"),
```
`-text` can also be used. For example:
```console
$ cat .gitattributes
*.tar.gz -text
$ printf "%s\n" test/data/* | git check-attr filter binary text --stdin
test/data/porg_internal-1.26.2.tar.gz: filter: unspecified
test/data/porg_internal-1.26.2.tar.gz: binary: unspecified
test/data/porg_internal-1.26.2.tar.gz: text: unset
test/data/porg_internal-1.26.3.tar.gz: filter: unspecified
test/data/porg_internal-1.26.3.tar.gz: binary: unspecified
test/data/porg_internal-1.26.3.tar.gz: text: unset
test/data/uv-test-berlin.lock: filter: unspecified
test/data/uv-test-berlin.lock: binary: unspecified
test/data/uv-test-berlin.lock: text: unspecified
```

```suggestion
            ("git", "check-attr", "filter", "binary", "text", "-z", "--stdin"),
```

rzuckerm commented on pre_commit_hooks/detect_non_ascii_characters.py:
```diff
@@ -0,0 +1,747 @@
+            filename, attr, value = stdout[i], stdout[i + 1], stdout[i + 2]
+            if (attr == "filter" and value == "lfs") or (
+                attr == "binary" and value == "set"
+            ):
```
```suggestion
            if (attr == "filter" and value == "lfs") or (
                attr == "binary" and value == "set"
            ) or (
                attr == "text" and value == "unset"
            ):
```

rzuckerm commented on tests/detect_non_ascii_characters_test.py:
```diff
@@ -0,0 +1,1402 @@
+            },
+            id="multiple_sub_directory_with same_file_name",
+        ),
+    ],
```
A test case needs to be added for `-text`

Commit:
Enhance binary file detection to include text attribute and add test case for -text files

rzuckerm reviewed (approved):
LGTM!

Commit:
Update pre-commit hook entries to version 5.0.0-2 for consistency and improvements

Commit:
add space to end of file
